arch/x86/kvm/pkvm/pgtable.c

```
@@ -462,3 +501,33 @@ int pkvm_pgtable_host_mmu_unmap(unsigned long vaddr, unsigned long size)
 
 	return pkvm_pgtable_walk(&host_mmu, vaddr, size, &walker);
 }
+
+void pkvm_pgtable_lookup(struct pkvm_pgtable *pgt, unsigned long vaddr,
+			 unsigned long *phys, u64 *prot, int *level)
+{
+	struct pkvm_pgtable_lookup_data data = {
+		.vaddr = vaddr,
+	};
+	struct pkvm_pgtable_walker walker = {
+		.cb = lookup_walker,
+		.arg = &data,
+		.walk_flags = PKVM_PGTABLE_WALK_LEAF,
+	};
+	int ret, retry_cnt = 0;
+
+retry:
+	ret = pkvm_pgtable_walk(pgt, vaddr, PAGE_SIZE, &walker);
+	/*
+	 * Give a few (e.g, 5 times) chances to re-walk pgtable if others change
+	 * the PTE during walking.
+	 */
+	if ((ret == -EAGAIN) && (retry_cnt++ < 5))
```

Looks like a hack, right? If all 5 retries fail, pkvm_pgtable_lookup() silently fails to update `phys`, `prot` and `level`, we have no idea about that so we're gonna use some old/random values of `phys`, `prot` and `level`.

---

Lookup walker will set the data with some default value, not random value. I can move them to this function to make it more clear.